README.md
aws-codedeploy-agent Cookbook
===================
Important Note: This cookbook is under development. Contribution is welcome!

Install AWS aws-codedeploy-agent

Requirements
------------
The following platforms are supported by this cookbook.

* Ubuntu 12.04/14.04/16.04
* CentOs 7 
* Fedora 21
* Debian 8 
* Amazon Linux AMI 2015.03

#### packages
- `aws-codedeploy-agent` - install aws-codedeploy-agent 

Attributes
----------
Customize the attributes to set ruby version requirements of AWS aws-codedeploy-agent (currently 2.0.0-p645) 

#### aws-codedeploy-agent::default

<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>node['aws-codedeploy-agent']['rbenv_ruby-version']</tt></td>
    <td>string</td>
    <td>Set the default ruby version of code deploy</td>
    <td><tt>2.1.5</tt></td>
  </tr>
</table>


Usage
-----
#### aws-codedeploy-agent::default

Just include `aws-codedeploy-agent` in your node's `run_list`:

```json
{
  "name":"my_node",
  "run_list": [
    "recipe[aws-codedeploy-agent]"
  ]
}
```

Testing
-------
Default .kitchen.yml test the cookbook in docker containers.

To test locally:

```bash
kitchen list
kitchen converge aws-codedeploy-agent-ubuntu-1604
kitchen verify aws-codedeploy-agent-ubuntu-1604
``` 

To test the cookbook on AWS instances use:

Configure the access to your AWS account using the AWS CLI:
```bash
aws configure --profile=AWS_CLI_PROFILE
```

copy the .kitchen-aws.yml to .kitchen-aws.local.yml and configure to suite your AWS account.

```bash
AWS_PROFILE=AWS_CLI_PROFILE KITCHEN_YAML=.kitchen-aws.local.yml bundle exec kitchen list 

AWS_PROFILE=AWS_CLI_PROFILE KITCHEN_YAML=.kitchen-aws.local.yml bundle exec kitchen verify aws-codedeploy-agent-ubuntu-1604-aws
AWS_PROFILE=AWS_CLI_PROFILE KITCHEN_YAML=.kitchen-aws.local.yml bundle exec kitchen verify aws-codedeploy-agent-ubuntu-1404-aws
AWS_PROFILE=AWS_CLI_PROFILE KITCHEN_YAML=.kitchen-aws.local.yml bundle exec kitchen verify aws-codedeploy-agent-ubuntu-1204-aws 
AWS_PROFILE=AWS_CLI_PROFILE KITCHEN_YAML=.kitchen-aws.local.yml bundle exec kitchen verify aws-codedeploy-agent-debian-8-aws 
AWS_PROFILE=AWS_CLI_PROFILE KITCHEN_YAML=.kitchen-aws.local.yml bundle exec kitchen verify aws-codedeploy-agent-centos-7-aws 
AWS_PROFILE=AWS_CLI_PROFILE KITCHEN_YAML=.kitchen-aws.local.yml bundle exec kitchen verify aws-codedeploy-agent-amazon-linux
```

To test all at once:

```bash
AWS_PROFILE=AWS_CLI_PROFILE KITCHEN_YAML=.kitchen-aws.local.yml bundle exec kitchen verify -c 5
```

Refer to the [kitchen-ec2](https://github.com/test-kitchen/kitchen-ec2) documentation

Contributing
------------
1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
Authors: 

* Oswald De Riemaecker (oswald@continuousphp.com)

```text
Copyright 2015, Continuous S.A.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
``` 


CHANGELOG.md
aws-codedeploy-agent CHANGELOG
====================

This file is used to list changes made in each version of the aws-codedeploy-agent cookbook.

0.6.0
-----
- Oswald De Riemaecker - Swithing to ruby_rbenv with rvm_download plugin to add a download installer to install ruby faster (except for centos)

0.4.0
-----
- Oswald De Riemaecker - Added Ubuntu 12.04 and Debian 8 Support
- Oswald De Riemaecker - Adding .foodcritic

0.3.0
-----
- Oswald De Riemaecker - Added awscli Support for onpremise support
- Oswald De Riemaecker - Split manual and package installer

0.2.0
-----
- Oswald De Riemaecker - Fixed deployment root folder

0.1.0
-----
- Oswald De Riemaecker - Initial release of aws-codedeploy-agent

- - -


recipes/default.rb
#
# Cookbook Name:: aws-codedeploy-agent
# Recipe:: default
#
# Copyright 2015, continuousphp.com
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

Chef::Log.debug('Platform:' + node['platform'])
Chef::Log.debug('Platform:' + node['platform_version'])

case node['platform']
when 'ubuntu'
  execute 'apt-get-update-periodic' do
    command 'apt-get update'
    ignore_failure true
  end

  case node['platform_version']
  when '16.04'
    %w(unzip rsync ruby2.3).each do |pkg|
      package pkg
    end
    download_installer
  when '14.04'
    %w(unzip rsync ruby2.0).each do |pkg|
      package pkg
    end
    download_installer
  when '12.04'
    %w(unzip rsync ruby).each do |pkg|
      package pkg
    end
    download_installer
  end
  service 'codedeploy-agent' do
    action [:enable, :start]
  end

when 'fedora'
  %w(unzip rsync ruby tar openssl-devel readline-devel zlib-devel initscripts).each do |pkg|
    package pkg
  end
  download_installer
  service 'codedeploy-agent' do
    action [:enable,:start]
    #provider Chef::Provider::Service::Init
  end

when 'centos'
  %w(unzip rsync ruby tar openssl-devel readline-devel zlib-devel bzip2 initscripts).each do |pkg|
    package pkg
  end
  manual_installer
  service 'codedeploy-agent' do
    action [:enable,:start]
    Chef::Provider::Service::Upstart
  end

when 'debian'
  execute 'apt-get-update-periodic' do
    command 'apt-get update'
    ignore_failure true
  end
  %w(unzip rsync ruby tar).each do |pkg|
    package pkg
  end
  download_installer
  service 'codedeploy-agent' do
    action [:enable,:start]
  end

when 'amazon'
  %w(ruby aws-cli).each do |pkg|
    package pkg
  end
  amazon_installer
  service 'codedeploy-agent' do
    action [:start]
    provider Chef::Provider::Service::Init
  end
else
  %w(unzip rsync ruby tar openssl-devel readline-devel zlib-devel).each do |pkg|
    package pkg
  end
  manual_installer
  service 'codedeploy-agent' do
    action [:enable,:start]
  end
end
